chia/full_node/coin_store.py

```diff
@@ -515,16 +518,27 @@ async def _set_spent(self, coin_names: List[bytes32], index: uint32):
         for coin_name in coin_names:
             r = self.coin_record_cache.get(coin_name)
             if r is not None:
+                if r.spent_block_index != uint32(0):
+                    raise ValueError(f"Coin already spent in cache: {coin_name}")
+
                 self.coin_record_cache.put(
                     r.name, CoinRecord(r.coin, r.confirmed_block_index, index, r.coinbase, r.timestamp)
                 )
             updates.append((index, self.maybe_to_hex(coin_name)))
 
-        if updates != []:
-            async with self.db_wrapper.write_db() as conn:
-                if self.db_wrapper.db_version == 2:
-                    await conn.executemany("UPDATE OR FAIL coin_record SET spent_index=? WHERE coin_name=?", updates)
-                else:
-                    await conn.executemany(
-                        "UPDATE OR FAIL coin_record SET spent=1,spent_index=? WHERE coin_name=?", updates
-                    )
+        assert len(updates) == len(coin_names)
+        async with self.db_wrapper.write_db() as conn:
+            if self.db_wrapper.db_version == 2:
+                ret: Cursor = await conn.executemany(
+                    "UPDATE OR FAIL coin_record SET spent_index=? WHERE coin_name=? AND spent_index=0", updates
```

Comment:
there is an (implicit) index for `coin_name` (since it's the primary key). We also have an index for `spent_index`. If sqlite decides to use the latter index, this operation will be very slow (because all unspent coins have `spent_index=0`, so you get a linear search among those).

I think it would be good to explicitly specify `INDEX BY` the correct index here, just to make sure sqlite and future readers of this code understands what the intention is. However, I don't know how to name the implicit primary key index.

Reply:
We had this problem before but I don't remember the conclusion. Would `index by coin_name` solve the performance issue?

Reply:
something *like* that, yes. but there is no index called `coin_name`. That index is created implicitly